src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -580,7 +698,166 @@ private static ImageHelper createImageProvider(JlinkConfiguration config,
+            // fall-through
+        }
+        return false;
+    }
+
+    private static InputStream getDiffInputStream() {
+        try {
+            String resourceName = String.format(DIFF_PATTERN, "java.base");
+            return JlinkTask.class.getModule().getResourceAsStream(resourceName);
+        } catch (IOException e) {
+            if (DEBUG) {
+                System.err.println("Failed to get diff pattern resource");
+                e.printStackTrace();
+            }
+            return null;
+        }
+    }
+
+    private static Archive newArchive(String module,
+                                      Path path,
+                                      Runtime.Version version,
+                                      boolean ignoreSigning,
+                                      JlinkConfiguration config) {
+        if (path.toString().endsWith(".jmod")) {
+            return new JmodArchive(module, path);
+        } else if (path.toString().endsWith(".jar")) {
+            ModularJarArchive modularJarArchive = new ModularJarArchive(module, path, version);
+            try (Stream<Archive.Entry> entries = modularJarArchive.entries()) {
+                boolean hasSignatures = entries.anyMatch((entry) -> {
+                    String name = entry.name().toUpperCase(Locale.ROOT);
+
+                    return name.startsWith("META-INF/") && name.indexOf('/', 9) == -1 && (
+                            name.endsWith(".SF") ||
+                                    name.endsWith(".DSA") ||
+                                    name.endsWith(".RSA") ||
+                                    name.endsWith(".EC") ||
+                                    name.startsWith("META-INF/SIG-")
+                    );
+                });
+
+                if (hasSignatures) {
+                    if (ignoreSigning) {
+                        System.err.println(taskHelper.getMessage("warn.signing", path));
+                    } else {
+                        throw new IllegalArgumentException(taskHelper.getMessage("err.signing", path));
+                    }
+                }
+            }
+            return modularJarArchive;
+        } else if (!"jrt".equals(path.toUri().getScheme()) && Files.isDirectory(path)) {
+            // The jrt URI path scheme conditional is there since we'd otherwise
+            // enter this branch for linkable JDK runtimes where the path is
+            // a jrt path and for the specific JDK module is a directory.
+            Path modInfoPath = path.resolve("module-info.class");
+            if (Files.isRegularFile(modInfoPath)) {
+                return new DirArchive(path, findModuleName(modInfoPath));
+            } else {
+                throw new IllegalArgumentException(
+                        taskHelper.getMessage("err.not.a.module.directory", path));
+            }
+        } else if (config.linkFromRuntimeImage()) {
+            // This is after all other archive types, since user-provided
```

Comment:
This block can be a utility method in the new `LinkableRuntimeImage` class as the logic can be encapsulated.